packages/web/src/components/Form.js

```diff
@@ -0,0 +1,191 @@
+// Module imports
+import {
+	createContext,
+	useCallback,
+	useContext,
+	useEffect,
+	useReducer,
+} from 'react'
+import PropTypes from 'prop-types'
+
+
+
+
+
+// Local imports
+import {
+	auth,
+	firestore,
+} from 'helpers/firebase'
+
+
+
+
+
+// Constants
+const INITIAL_STATE = {
+	errors: {},
+	isValid: false,
+	isTouched: false,
+	touched: {},
+	validity: {},
+	values: {},
+}
+
+
+
+
+
+function reducer(state, action) {
```

Comment:
Is there a reason we are defining our own reducer function instead of using the built-in one provided by React?

Reply:
What reducer are you referring to? React doesn't provide any default reducer, though it does provide [`useReducer`](https://reactjs.org/docs/hooks-reference.html#usereducer) (which we're using.